File: bvp/stencil_driver.cpp
/* Main routine to run a stencil-based finite-difference method for solution of a
   second-order, scalar-valued BVP:

      u'' = p(t)*u' + q(t)*u + r(t),  a<t<b,
      u(a) = ua,  u(b) = ub

   where the problem has stiffness that may be adjusted using
   the real-valued parameter lambda<0 [read from the command line]


   This driver attempts to solve the problem using a second-order, stencil-based
   finite-difference approximation.

   D.R. Reynolds
   Math 6321 @ SMU
   Fall 2020  */

#include <iostream>
#include <iomanip>
#include <vector>
#include <armadillo>
#include "bvp.hpp"

using namespace std;
using namespace arma;

#ifndef USE_SPARSE
#define SPARSE_LINALG 0
#else
#define SPARSE_LINALG 1
#endif

// main routine
int main(int argc, char **argv) {

  // get lambda from the command line, otherwise set to -10
  double lambda;
  if (argc > 1)
    lambda = atof(argv[1]);
  if (lambda >= 0.0)
    lambda = -10.0;

  // define BVP object
  BVP bvp(lambda);

  // loop over spatial resolutions for tests
  vector<int> N = {100, 1000, 10000};
  for (int i=0; i<N.size(); i++) {

    // output problem information
    cout << "\nStencil-based FD method for BVP with lambda = "
         << lambda << ":" << ",  N = " << N[i] << endl;

    // compute/store analytical solution
    vec t = linspace(bvp.a, bvp.b, N[i]+1);
    double h = t(1)-t(0);
    vec utrue(N[i]+1);
    for (int j=0; j<=N[i]; j++)
      utrue(j) = bvp.utrue(t(j));

    // create matrix and right-hand side vectors
#ifdef SPARSE_LINALG
    sp_mat A(N[i]+1,N[i]+1);
#else
    mat A(N[i]+1,N[i]+1);
    A.fill(0.0);
#endif
    vec b(N[i]+1);
    b.fill(0.0);

    // set up linear system
    A(0,0) = 1.0;
    b(0) = bvp.ua;
    A(N[i],N[i]) = 1.0;
    b(N[i]) = bvp.ub;
    for (int j=1; j<N[i]; j++) {
      A(j,j-1) = -1.0 - 0.5*h*bvp.p(t(j));
      A(j,j) = 2.0 + h*h*bvp.q(t(j));
      A(j,j+1) = -1.0 + 0.5*h*bvp.p(t(j));
      b(j) = -h*h*bvp.r(t(j));
    }

    // solve linear system for BVP solution
#ifdef SPARSE_LINALG
    vec u = arma::spsolve(A,b);
#else
    vec u = arma::solve(A,b);
#endif

    // output maximum error
    vec uerr = abs(u-utrue);
    cout << "  Maximum BVP solution error = " << std::setprecision(4)
       << uerr.max() << "\n";
  }

  return 0;
}

File: bvp/stencil_driver.py
#!/usr/bin/env python3
#
# Main routine to run a stencil-based finite-difference method for solution of a
# second-order, scalar-valued BVP:
#
#    u'' = p(t)*u' + q(t)*u + r(t),  a<t<b,
#    u(a) = ua,  u(b) = ub
#
# where the problem has stiffness that may be adjusted using
# the real-valued parameter lambda<0 [read from the command line]
#
# This driver attempts to solve the problem using a second-order, stencil-based
# finite-difference approximation.
#
# D.R. Reynolds
# Math 6321 @ SMU
# Fall 2023

import sys
import numpy as np
import scipy.sparse as sp
from scipy.sparse.linalg import spsolve
from BVP import *

# get lambda from the command line, otherwise set to -10
lam = -10.0
if (len(sys.argv) > 1):
    lam = float(sys.argv[1])

# create BVP object
bvp = BVP(lam)

# loop over spatial resolutions for tests
N = [100, 1000, 10000]
for n in N:

    # output problem information
    print("\nStencil-based FD method for BVP with lambda =", lam, ",  N =", n)

    # compute/store analytical solution
    t = np.linspace(bvp.a, bvp.b, n+1)
    h = t[1]-t[0]
    utrue = np.zeros(n+1)
    for j in range(n+1):
      utrue[j] = bvp.utrue(t[j])

    # create matrix and right-hand side vectors
    Arows = np.zeros(3*(n+1)-4)
    Acols = np.zeros(3*(n+1)-4)
    Avals = np.zeros(3*(n+1)-4)
    b = np.zeros(n+1)

    # set up linear system
    idx = 0
    Arows[idx] = 0   # A(0,0) = 1.0
    Acols[idx] = 0
    Avals[idx] = 1
    idx += 1
    b[0] = bvp.ua

    Arows[idx] = n
    Acols[idx] = n
    Avals[idx] = 1   # A(n,n) = 1.0
    idx += 1
    b[n] = bvp.ub
    for j in range(1,n):
        pj = bvp.p(t[j])
        qj = bvp.q(t[j])

        Arows[idx] = j     # A[j,j-1]
        Acols[idx] = j-1
        Avals[idx] = -1 - 0.5*h*pj
        idx += 1
        Arows[idx] = j     # A[j,j]
        Acols[idx] = j
        Avals[idx] = 2 + h*h*qj
        idx += 1
        Arows[idx] = j     # A[j,j+1]
        Acols[idx] = j+1
        Avals[idx] = -1 + 0.5*h*pj
        idx += 1
        b[j] = -h*h*bvp.r(t[j])
    A = sp.csr_matrix(sp.coo_matrix((Avals, (Arows, Acols)), shape=(n+1, n+1)))

    # solve linear system for BVP solution
    u = spsolve(A,b)

    # output maximum error
    uerr = np.abs(u-utrue)
    print("  Maximum BVP solution error = %.4e" % (np.max(uerr)))
